### PR TITLE
Fix eval_accumulation_steps leading to incorrect metrics

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Uses the logic in the `GradientState` to know when we've reached the end of training and should sync the gradients. 

Doing so relies on [this](https://github.com/huggingface/accelerate/blob/main/src/accelerate/accelerator.py#L862-L869) code, which already checks for the case of if a dataloader has no length and works properly

Fixes # (issue)
Solves https://github.com/huggingface/transformers/issues/24734

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Nice!
